### PR TITLE
add qg example

### PR DESCRIPTION
**Pull request by westfish**
### PR types

New features

### PR changes

Models

### Description

增加问题生成example，包括基于英语预训练语言模型t5和基于中文预训练语言模型unimo-text两个子文件夹，其中unimo-text除训练预测评估外，还包括模型加速推理和部署。目前unimo-text下提供的模型参数还没有包括在大数据集(自建的9个领域的合数据集)上进行微调的参数。

### CODE REVIEW

File: examples/question_generation/README.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# 问题生成
+
+Question Generation（QG），即问题生成，指的是给定一段上下文和答案，自动生成一个流畅且符合上下文主题的问句。问题生成技术在教育、咨询、搜索、问答等多个领域均有着巨大的应用价值。
+
+PaddleNLP提供英文和中文问题生成任务示例，分别基于英文预训练语言模型[t5](./t5)和中文预训练语言模型[unimo-text](./unimo-text)。
```
Comment (wawltor) [inline examples/question_generation/README.md:5]: 建议在这里增加一些相关论文的引用的篇幅，让引入的页面丰富一点

同时在中文和英文可以按照列表方式展开

File: examples/question_generation/t5/README.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+### 数据准备
+
+#### 数据加载
+**SQuAD**（Stanford Question Answering Dataset）数据集是一个英文问答数据集，现有的问题生成研究主要在该数据集上进行评价。**SQuAD**中的数据由段落、问题、答案3个主要部分组成，其中段落从维基百科中获取，问题和答案通过众包的方式由人工标注。
+
+为了方便用户快速测试，PaddleNLP Dataset API内置了Squad数据集，一键即可完成数据集加载，示例代码如下：
+
+```python
+from paddlenlp.datasets import load_dataset
+train_set, dev_set, test_set = load_dataset("squad",  splits=["train_v1", "dev_v1"])
```
Comment (wawltor) [inline examples/question_generation/t5/README.md:41]: 这里的SQuAD数据为什么会有v1的版本了？

File: examples/question_generation/t5/README.md [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+具体案例如下，
+```text
+answer: the Miller–Rabin primality test context: The property of being prime (or not) is called primality. A simple but slow method of verifying the primality of a given number n is known as trial division. It consists of testing whether n is a multiple of any integer between 2 and . Algorithms much more efficient than trial division have been devised to test the primality of large numbers. These include the Miller–Rabin primality test, which is fast but has a small probability of error, and the AKS primality test, which always produces the correct answer in polynomial time but is too slow to be practical. Particularly fast methods are available for numbers of special forms, such as Mersenne numbers. As of January 2016[update], the largest known prime number has 22,338,618 decimal digits.
+
+question: What is the name of the process which confirms the primality of a number n?
+```
+
+### 模型训练
+
+运行如下命令即可在训练集上进行finetune，并在验证集上进行验证
```
Comment (wawltor) [inline examples/question_generation/t5/README.md:59]: 建议按照标准改一下名字

File: examples/question_generation/t5/README.md [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+- `warmup_proportion` 表示warmup_steps所占总步数的比例。学习率逐渐升高到基础学习率（即上面配置的learning_rate）所需要的迭代数。
+
+- `device` 表示使用的设备。
+
+程序运行时将会自动进行训练和验证，训练过程中会自动保存模型在指定的`output_dir`中。如：
+
+```text
+./output/
+├── t5_model_1000.pdparams
```
Comment (wawltor) [inline examples/question_generation/t5/README.md]: output的模型路径建议 pdparams后缀去掉，因为pdparams指代的是动态图参数，但是这个目录还有一些tokenizer存储

File: examples/question_generation/t5/README.md [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+
+### 模型预测
+
+运行如下命令即可在验证集上进行测试
+
+```shell
+# GPU启动，预测仅支持单卡
+export CUDA_VISIBLE_DEVICES=0
+python generate.py \
+    --model_name_or_path=t5-base-finetuned-question-generation-ap \
```
Comment (wawltor) [inline examples/question_generation/t5/README.md]: 这里建议直接写成一个目录？这里可能会误导用户直接使用了微调后的目录

File: examples/question_generation/t5/finetune.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+import paddle.nn as nn
+from paddle.io import BatchSampler, DistributedBatchSampler, DataLoader
+from paddlenlp.transformers import T5ForConditionalGeneration, T5Tokenizer
+from paddlenlp.transformers import LinearDecayWithWarmup
+from paddlenlp.utils.log import logger
+from paddlenlp.datasets import load_dataset
+from paddlenlp.data import Tuple, Stack, Pad
+from utils import convert_example, compute_metrics
+
+
```
Comment (wawltor) [inline examples/question_generation/t5/finetune.py]: 这里使用 yapf: disable  来处理一下，和unimo-text模型保持一致

File: examples/question_generation/t5/finetune.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+    # Generate parameter names needed to perform weight decay.
+    # All bias and LayerNorm parameters are excluded.
+    decay_params = [
+        p.name for n, p in model.named_parameters()
+        if not any(nd in n for nd in ["bias", "norm"])
+    ]
+    optimizer = paddle.optimizer.AdamW(
+        learning_rate=lr_scheduler,
+        beta1=0.9,
```
Comment (wawltor) [inline examples/question_generation/t5/finetune.py]: beta1 和 beta2 这两个看起来不用写成两个magic number，默认就是 0.9 0.999

File: examples/question_generation/t5/finetune.py [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+        beta2=0.999,
+        epsilon=args.adam_epsilon,
+        parameters=model.parameters(),
+        weight_decay=args.weight_decay,
+        apply_decay_param_fun=lambda x: x in decay_params)
+
+    if args.use_amp:
+        scaler = paddle.amp.GradScaler(init_loss_scaling=args.scale_loss)
```
Comment (wawltor) [inline examples/question_generation/t5/finetune.py]: 这里的AMP看起来只是提供了O1的方式，看看是否要提供O2的方式

File: examples/question_generation/t5/finetune.py [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+                    if not os.path.exists(output_dir):
+                        os.makedirs(output_dir)
+                    # Need better way to get inner model of DataParallel
+                    model_to_save = model._layers if isinstance(
+                        model, paddle.DataParallel) else model
+                    model_to_save.save_pretrained(output_dir)
+                    tokenizer.save_pretrained(output_dir)
+            if global_step >= num_training_steps:
+                return
+    if paddle.distributed.get_rank() == 0:
```
Comment (wawltor) [inline examples/question_generation/t5/finetune.py]: 这里的保存逻辑最好按照best model来存储，而不是final存储

File: examples/question_generation/t5/finetune_run.sh [hunk 10]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (wawltor) [inline examples/question_generation/t5/finetune_run.sh]: 去掉

File: examples/question_generation/t5/generate_run.sh [hunk 11]
```diff
@@ -0,0 +1,29 @@
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (wawltor) [inline examples/question_generation/t5/generate_run.sh]: 去掉

File: examples/question_generation/t5/generate.py [hunk 12]
```diff
@@ -0,0 +1,240 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (wawltor) [inline examples/question_generation/t5/generate.py]: 2021-> 2022

File: examples/question_generation/t5/finetune.py [hunk 13]
```diff
@@ -0,0 +1,324 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
```
Comment (wawltor) [inline examples/question_generation/t5/finetune.py]: 2021->2022

File: examples/question_generation/t5/generate.py [hunk 14]
```diff
@@ -0,0 +1,240 @@
...
+import paddle
+from paddle.io import BatchSampler, DataLoader
+from paddlenlp.datasets import load_dataset
+from paddlenlp.data import Tuple, Stack, Pad
+from paddlenlp.transformers import T5ForConditionalGeneration, T5Tokenizer
+from utils import convert_example, compute_metrics
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
```
Comment (wawltor) [inline examples/question_generation/t5/generate.py]: 同上 yapf disable一下

File: examples/question_generation/t5/utils.py [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import numpy as np
+import nltk
+from paddlenlp.metrics import BLEU
+import evaluate
```
Comment (wawltor) [inline examples/question_generation/t5/utils.py:16]: 第三方包提前

File: examples/question_generation/t5/utils.py [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+        print(example['context'])
+        print(example['question'])
+        print(example['answers'])
+        print(example['answer_starts'])
+        print(example['is_impossible'])
+
+    input_seq = f'answer: {answer} context: {context} </s>'
+    output_seq = f'question: {question} </s>'
+
+    labels = tokenizer(
```
Comment (wawltor) [inline examples/question_generation/t5/utils.py]: 这个命名是不是output比较好？ 同下input_ids

File: examples/question_generation/t5/utils.py [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+        output_seq,
+        max_seq_len=max_target_length,
+        pad_to_max_seq_len=True,
+        truncation_strategy="longest_first",
+    )
+
+    output_ids = [decoder_start_token_id] + labels["input_ids"][:-1]
+
+    if ignore_pad_token_for_loss:
```
Comment (wawltor) [inline examples/question_generation/t5/utils.py:58]: -100 这种magic number最好说明一下，或者看看能不能用tokenizer的speical id 来代替

File: examples/question_generation/unimo-text/README.md [hunk 18]
```diff
@@ -0,0 +1,309 @@
...
+
+PaddleNLP提供了方便易用的接口，可指定模型名或模型参数文件路径通过from_pretrained()方法加载不同网络结构的预训练模型，且相应预训练模型权重下载速度快速、稳定。
+Transformer预训练模型汇总包含了如 ERNIE、BERT、T5、UNIMO等主流预训练模型。下面以中文unimo-text-1.0模型为例，演示如何加载预训练模型和分词器：
+```
+from paddlenlp.transformers import  ErnieForGeneration, ErnieTokenizer
+model_name = "ernie-1.0"
+model = UNIMOLMHeadModel.from_pretrained(model_name)
+tokenizer = UNIMOTokenizer.from_pretrained(model_name)
+```
+<!--
```
Comment (wawltor) [inline examples/question_generation/unimo-text/README.md:43]: 这部分可以先去掉

File: examples/question_generation/unimo-text/README.md [hunk 19]
```diff
@@ -0,0 +1,309 @@
...
+├── train.py # 训练评估脚本
+├── utils.py # 工具函数脚本
+└── README.md # 说明文档
+```
+
+### 问题生成定制训练全流程介绍
+接下来，我们将按数据准备、训练、预测、推理部署等四个阶段对问题生成应用的全流程进行介绍。
+1. **数据准备**
+- 如果没有已标注的数据集，我们推荐doccano数据标注工具([doccano](https://github.com/doccano/doccano))。
+- 如果已有标注好的本地数据集，我们需要根据将数据集整理为文档要求的格式，请参考[从本地文件创建数据集](###从本地文件创建数据集)。
```
Comment (wawltor) [inline examples/question_generation/unimo-text/README.md]: 这里要处理一下，咱们这个是没有提到docann怎么处理数据的，直接说一下数据集来源即可

File: examples/question_generation/unimo-text/README.md [hunk 20]
```diff
@@ -0,0 +1,309 @@
...
+- `warmup_propotion` 表示学习率逐渐升高到基础学习率（即上面配置的learning_rate）所需要的迭代数占总步数的比例。
+- `max_seq_len` 模型输入序列的最大长度。
+- `max_target_len` 模型训练时标签的最大长度。
+- `min_dec_len` 模型生成序列的最小长度。
+- `max_dec_len` 模型生成序列的最大长度。
+- `do_train` 是否进行训练。
+- `do_predict` 是否进行预测，在验证集上会自动评估。
+- `device` 表示使用的设备，从gpu和cpu中选择。
+- `adversarial_training` 表示使用何种对抗训练策略，从['None', 'fgm', 'pgd']中选择。
+- `template` 表示使用的设备，从[0, 1, 2, 3]中选择，0表示不选择模版，1表示使用默认模版。
```
Comment (wawltor) [inline examples/question_generation/unimo-text/README.md]: 表示使用的设备 -> 表示使用的模板？

File: examples/question_generation/unimo-text/run_gen.py [hunk 21]
```diff
@@ -0,0 +1,302 @@
...
+import paddle.nn.functional as F
+from paddlenlp.transformers import LinearDecayWithWarmup
+from paddle.optimizer import AdamW
+
+from paddlenlp.datasets import load_dataset
+from paddlenlp.transformers import UNIMOLMHeadModel, UNIMOTokenizer, BasicTokenizer
+from paddlenlp.metrics import BLEU
+
+from gen_utils import print_args, set_seed, create_data_loader, select_sum
+from adversarial_utils import FGM, PGD
```
Comment (wawltor) [inline examples/question_generation/unimo-text/run_gen.py]: adversarial_utils 没有提交上来？

File: examples/question_generation/README.md (hunk 1 above)
Comment (westfish) [reply examples/question_generation/README.md:5]: 好的

File: examples/question_generation/t5/README.md (hunk 2 above)
Comment (westfish) [reply examples/question_generation/t5/README.md:41]: SQuAD有v1和v2两个版本，这里使用的v1版本

File: examples/question_generation/t5/README.md (hunk 3 above)
Comment (westfish) [reply examples/question_generation/t5/README.md:59]: 好的

File: examples/question_generation/t5/README.md (hunk 4 above)
Comment (westfish) [reply examples/question_generation/t5/README.md]: 好的👌

File: examples/question_generation/t5/README.md (hunk 5 above)
Comment (westfish) [reply examples/question_generation/t5/README.md]: 有道理 好的

File: examples/question_generation/unimo-text/export_model.py [hunk 22]
```diff
@@ -0,0 +1,142 @@
...
+from pprint import pprint
+
+from paddlenlp.transformers import UNIMOLMHeadModel, UNIMOTokenizer
+from paddlenlp.ops import FasterUNIMOText
+
+from paddlenlp.utils.log import logger
+
+
+def parse_args():
+    parser = argparse.ArgumentParser()
```
Comment (wawltor) [inline examples/question_generation/unimo-text/export_model.py:30]: 这里建议还是yapf disable一下

File: examples/question_generation/unimo-text/export_model.py [hunk 23]
```diff
@@ -0,0 +1,142 @@
...
+    parser.add_argument("--length_penalty",
+                        default=1.2,
+                        type=float,
+                        help="The diversity rate to procedure beam search. ")
+    args = parser.parse_args()
+    return args
+
+
+def do_predict(args):
+    place = "gpu"
```
Comment (wawltor) [inline examples/question_generation/unimo-text/export_model.py:49]: 这里的place建议不需要用gpu，因为要考虑没有GPU的环境，直接cpu或者默认device即可

File: examples/question_generation/unimo-text/export_model.py [hunk 24]
```diff
@@ -0,0 +1,142 @@
...
+                                 trans_out=True)
+
+    # Set evaluate mode
+    unimo_text.eval()
+
+    # Convert dygraph model to static graph model
+    unimo_text = paddle.jit.to_static(
+        unimo_text,
+        input_spec=[
+            # input_ids
```
Comment (wawltor) [inline examples/question_generation/unimo-text/export_model.py:67]: 考虑到windows兼容的问题，这里建议使用int64

File: examples/question_generation/unimo-text/predict.py [hunk 25]
```diff
@@ -0,0 +1,301 @@
+# Copyright (c) 2022 PaddlePaddle Authors. All Rights Reserved.
```
Comment (wawltor) [inline examples/question_generation/unimo-text/predict.py:1]: 看起来是train.py和predict.py有点混了，train.py变成一个evaluate的脚本 ，infer.py变成了一个train和evaluate的过程

File: examples/question_generation/unimo-text/gen_utils.py [hunk 26]
```diff
@@ -0,0 +1,322 @@
...
+                                                 max_seq_len=max_seq_len,
+                                                 max_target_len=max_target_len,
+                                                 max_title_len=max_title_len,
+                                                 return_position_ids=True,
+                                                 return_length=True)
+        temp_tokens = tokenizer.convert_ids_to_tokens(
+            tokenized_example['input_ids'])
+        index_list = []
+        count = tokenized_example['input_ids'].count(tokenizer.cls_token_id)
+        assert count == 7 or count == 2, str(
```
Comment (wawltor) [inline examples/question_generation/unimo-text/gen_utils.py:116]: 这里需要解释一下 为什么 cls_token_id token为2，7的时候  要assert

**Review by wawltor — APPROVED** [review]
LGTM
